### PR TITLE
Release 1.0.1 changes

### PR DESCRIPTION
**Pull request by paweld2**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,7 +62,11 @@ in the following manner:
 CastleApi newAPIRef = Castle.sdk().onRequest(req)
 ```
 
-Here `req` is an instance of `HttpServletRequest`.
+Here `req` is an instance of `HttpServletRequest`. 
+
+Note that the `req` instance should be bind to the underlaing request in order to extract the necessary information. 
+It means that a safe place to create the `CastleApi` instance is the request handling thread. After creation the 
+`CastleApi` instance can be passed to any thread independently of the original thread life cycle.
```
Comment (paweld2) [inline README.md]: Additional information about async use of the CastleApi. I do not know if we should be more explicit here.

File: src/main/java/io/castle/client/api/CastleApi.java [hunk 2]
```diff
@@ -130,22 +130,35 @@
...
     /**
+     * Makes an async POST request to the track endpoint containing required and optional parameters.
+     *
+     * @param event      a String representing an event understood by the Castle API
+     * @param userId     a String representing a user ID
+     * @param properties object for recording additional information connected to the event, takes null
+     * @param trait      object for recording additional information about the user like email or name, takes null
+     * @see <a href="https://api.castle.io/docs#track">The docs</a>
+     */
+    void track(String event, @Nullable String userId, @Nullable Object properties, @Nullable Object trait);
```
Comment (paweld2) [inline src/main/java/io/castle/client/api/CastleApi.java]: The missing trait parameter.

File: src/main/java/io/castle/client/internal/utils/CastleContextBuilder.java [hunk 3]
```diff
@@ -65,8 +65,18 @@ private void addHeaderValue(ArrayList<CastleHeader> castleHeadersList, String ke
...
+     * Extract the clientId from the request.
+     * If __cid cookie setup the use that value, if not use header 'X-Castle-Client-Id', if not use empty string ''
+     * @param request
+     * @return
+     */
     private String setClientIdFromHttpServletRequest(HttpServletRequest request) {
         String cid = request.getHeader("X-Castle-Client-Id");
+        if( cid == null) {
+            //Default value is empty string, so client_id always will be send.
+            cid = "";
```
Comment (paweld2) [inline src/main/java/io/castle/client/internal/utils/CastleContextBuilder.java]: This ensure that the client_id will be "" if no value is found. Tests check this.

File: src/test/java/io/castle/client/CastleTrackHttpTest.java [hunk 4]
```diff
@@ -59,7 +59,35 @@ public void trackEndpointWithUserIDAndPropertiesTest() throws InterruptedExcepti
...
+        CustomAppTraits trait = new CustomAppTraits();
+        trait.setX("x value");
+        trait.setY(2342);
+
+        // and an track request is made
+        sdk.onRequest(request).track(event, id, properties, trait);
+
+        // then
+        RecordedRequest recordedRequest = server.takeRequest();
+        Assert.assertEquals("{\"name\":\"$login.succeeded\",\"user_id\":\"23456\",\"context\":{\"active\":true,\"client_id\":\"\",\"ip\":\"127.0.0.1\",\"headers\":{\"REMOTE_ADDR\":\"127.0.0.1\"},\"library\":{\"name\":\"Castle\",\"version\":\"1.0.1\"}},\"properties\":{\"a\":\"valueA\",\"b\":123456},\"trait\":{\"x\":\"x value\",\"y\":2342}}",
```
Comment (paweld2) [inline src/test/java/io/castle/client/CastleTrackHttpTest.java]: Tests that check that the trait parameter is send.

File: README.md [hunk 5]
```diff
@@ -53,7 +62,11 @@ in the following manner:
 CastleApi newAPIRef = Castle.sdk().onRequest(req)
 ```
 
-Here `req` is an instance of `HttpServletRequest`.
+Here `req` is an instance of `HttpServletRequest`. 
+
+Note that the `req` instance should be bind to the underlaing request in order to extract the necessary information. 
+It means that a safe place to create the `CastleApi` instance is the request handling thread. After creation the
```
Comment (lluft) [inline README.md]: "bound to the underlying "

Comment (paweld2) [reply README.md]: fixed.

File: src/main/java/io/castle/client/api/CastleApi.java [hunk 6]
```diff
@@ -118,47 +118,59 @@
...
      * @see <a href="https://api.castle.io/docs#track">The docs</a>
      */
     void track(String event, @Nullable String userId);
 
     /**
+     * Makes an async POST request to the track endpoint containing all required parameters and a user ID.
+     *
+     * @param event    a String representing an event understood by the Castle API
+     * @param userId   a String representing a user ID
+     * @param reviewId a String representing a reference to review ID
```
Comment (paweld2) [inline src/main/java/io/castle/client/api/CastleApi.java]: The additional review_id parameter.

File: src/main/java/io/castle/client/internal/backend/OkRestApiBackend.java [hunk 7]
```diff
@@ -38,14 +38,19 @@ public OkRestApiBackend(OkHttpClient client, CastleGsonModel model, CastleConfig
...
         if (userId == null) {
             json.add("user_id", JsonNull.INSTANCE);
         } else {
             json.add("user_id", new JsonPrimitive(userId));
         }
+        if (reviewId == null) {
+            json.add("review_id", JsonNull.INSTANCE);
+        } else {
+            json.add("review_id", new JsonPrimitive(reviewId));
+        }
```
Comment (paweld2) [inline src/main/java/io/castle/client/internal/backend/OkRestApiBackend.java]: review_id added at the same level that user_id.

File: src/test/java/io/castle/client/CastleTrackHttpTest.java [hunk 8]
```diff
@@ -55,19 +76,20 @@ public void trackEndpointWithUserIDAndPropertiesTest() throws InterruptedExcepti
         properties.setB(123456);
 
         // and an track request is made
-        sdk.onRequest(request).track(event, id, properties);
+        sdk.onRequest(request).track(event, userId, reviewId, properties);
 
         // then
         RecordedRequest recordedRequest = server.takeRequest();
-        Assert.assertEquals("{\"name\":\"$login.succeeded\",\"user_id\":\"12345\",\"context\":{\"active\":true,\"client_id\":\"\",\"ip\":\"127.0.0.1\",\"headers\":{\"REMOTE_ADDR\":\"127.0.0.1\"},\"library\":{\"name\":\"Castle\",\"version\":\"1.0.1\"}},\"properties\":{\"a\":\"valueA\",\"b\":123456}}",
+        Assert.assertEquals("{\"name\":\"$login.succeeded\",\"user_id\":\"12345\",\"review_id\":\"r987\",\"context\":{\"active\":true,\"client_id\":\"\",\"ip\":\"127.0.0.1\",\"headers\":{\"REMOTE_ADDR\":\"127.0.0.1\"},\"library\":{\"name\":\"Castle\",\"version\":\"1.0.1\"}},\"properties\":{\"a\":\"valueA\",\"b\":123456}}",
```
Comment (paweld2) [inline src/test/java/io/castle/client/CastleTrackHttpTest.java]: Tests confirm that the parameter is send. We have 100% coverage as before.

**Pull request merged**